### ISSUE TITLE
Extending components and extracting common CSS

### ISSUE THREAD

**Issue by tanem** [suitcss/suit#106]
Hey all, hoping I can get some opinions on the below please?

The following code is a simplified version of something we've encountered recently. `FormOne-button` and `FormTwo-button` are extending the nested `Button` component:

```css
/* button.css */
.Button {
  width: 100px;
  font-size: 12px;
  font-weight: bold;
  background-color: white;
}

/* formOne.css */
.FormOne-button {
  font-size: 14px;
  font-weight: normal;
  background-color: blue;
}

/* formTwo.css */
.FormTwo-button {
  font-size: 16px;
  font-weight: normal; 
  background-color: blue;
}
```

```html
<div class="FormOne">
  <button class="FormOne-button Button" type="button">Button</button>
</div>

<div class="FormTwo">
  <button class="FormTwo-button Button" type="button">Button</button>
</div>
```

There is some common code in the buttons, so how could we extract that out? Should we extract a `FormButton` component, something like the below:

```css
/* formButton.css */
.FormButton {
  font-weight: normal;
  background-color: blue;
}

/* formOne.css */
.FormOne-button {
  font-size: 14px;
}

/* formTwo.css */
.FormTwo-button {
  font-size: 16px;
}
```

```html
<div class="FormOne">
  <button class="FormOne-button FormButton Button" type="button">Button</button>
</div>

<div class="FormTwo">
  <button class="FormTwo-button FormButton Button" type="button">Button</button>
</div>
```

Is stacking components like that (`FormButton` + `Button`) onto the one element recommended though? I suppose we could just have a `FormButton` component that contains all the `Button` styles we want, simplifying the HTML, but that just moves the duplication up a level.

**Comment by racse1** [comment]
Hello,

I worked around that topic recently. It's tough one if you don't use some kind of CSS processor. 

Personally I use Sass and your code will look like this:

```css
/* _button.scss */
.Button {
  width: 100px;
  font-size: 12px;
  font-weight: bold;
  background-color: white;
}

/* _form-button.scss */
@import 'button';

.FormButton {
  @extend .Button;

  font-weight: normal;
  background-color: blue;
}

.FormButton--smaller {
  font-size: 14px;
}

.FormButton--bigger {
  font-size: 16px;
}
```

You might be aware that Sass's `@extend` feature will generate the following code:

```css
.Button,
.FormButton {
  width: 100px;
  font-size: 12px;
  font-weight: bold;
  background-color: white;
}

.FormButton {
  font-weight: normal;
  background-color: blue;
}
```

Also, I recommend these wonderful articles about extending components with SUIT CSS:

* [Extending Styles](http://philipwalton.com/articles/extending-styles/) by @philipwalton
* [Nesting Components](http://simurai.com/blog/2015/05/11/nesting-components/) by @simurai

I hope we will get a couple more opinions here.

Thanks!

**Comment by simonsmith** [comment]
I think I would go the route of using a modifier for your common button form styles. So:

``` css
.Button {
  width: 100px;
  font-size: 12px;
  font-weight: bold;
  background-color: white;
}

.Button--form {
  font-weight: normal;
  background-color: blue;
}
```

You can add further modifiers to that component and compose variants of your buttons in the HTML.

@racse1 has linked to the excellent blog post by Philip Walton in which he highlights a great way of extending styles with an `import` and `extend`.

It depends on your tooling but I would use [postcss-extend](https://github.com/travco/postcss-extend) and [postcss-import](https://github.com/postcss/postcss-import) to achieve it.

**Comment by tanem** [comment]
Thanks @racse1 & @simonsmith :smile: The links were great reading.

To give you some background, we made a deliberate move to reduce our SASS usage to variables and a few mixins. This was because on our team, the distance it was creating from the outputted CSS was causing more issues than the benefits it was providing (similar to the reasons in [this article](https://medium.com/@fat/mediums-css-is-actually-pretty-fucking-good-b8e2a6c78b06)). For that reason we're a little reluctant to use `@extend` at the moment.

@simonsmith we thought very seriously about your suggestion of using modifiers on the core `Button` component, a la [this comment](http://simurai.com/blog/2015/05/11/nesting-components/#comment-2041894660) in [Nesting Components](http://simurai.com/blog/2015/05/11/nesting-components/). It feels pretty simple.

One thing that put us off at the time is the fact that a component's styles would then be spread throughout various places in the CSS architecture, which might make it harder to unpick/modify. Do you think this could be mitigated by a naming convention?

e.g. if we had two components named `FeedbackForm` and `PaymentForm`, and there were two button variations in `FeedbackForm`:

```css
/* button.css */
.Button {}
.Button--feedbackFormHappy {}
.Button--feedbackFormSad {}
.Button--paymentForm {}
```

I'm unsure about this approach though, since it seems to deviate from [what the docs recommend](https://github.com/suitcss/suit/blob/master/doc/components.md#styling-dependencies)?

**Comment by simonsmith** [comment]
In your example you highlighted common styles found in `FormButton` that you would use in both/all your forms. It's those styles that I would put into the `Button` component as they are generic and reusable.

Specific things for each form would use rules specific to that component, so:

``` html
<div class="FormOne">
  <button class="FormOne-btn Button Button--form" type="button">Button</button>
</div>

<div class="FormTwo">
  <button class="FormOne-btn Button Button--form" type="button">Button</button>
</div>
```

A minor change, but that's how I might approach it. I would agree that the `Button` component cannot be responsible for everything, just reusable stuff.

**Comment by tanem** [comment]
@simonsmith sorry, I misread your original post. I like the simplicity of your suggestion. It also keeps things closer to what I've gleaned from the docs.

If we used the multiple component approach, I was worried that there might be an explosion of intermediate components across the app. It also wasn't clear in my mind when we'd want to move styles up a level in a component hierarchy, instead of just creating another intermediate component. Guess that might have become clearer over time.

Anyways, will have a play around with reusable styles on `Button`. Thanks for your help :+1:
